### PR TITLE
Members endpoints with GET, POST, and PUT

### LINKED ISSUE #2: View all Members
- GET /members

### LINKED ISSUE #4: View Member Info
GET /members/:memberId

### LINKED ISSUE #5: Edit Member Info
- PUT /member/:memberId
- Needs Auth (logged in & role protection)
- Member can only edit their own fields (A-K)
- Director can edit everything

### PR DESCRIPTION
**Pull request by Qrtn**
## Summary

Endpoint to get and update data for members. Needed for the main table and individual member profile page

## Requests / Responses 
**Request**
GET `/api/members` Returns a list of members. Currently, fields like `areDuesPaid` and `level` are excluded if the user is not at director level.

**Response**
HTTP/1.1 200 OK

```
{
  "success": true,
  "result": [
    {
      "firstName": "Jeffrey",
      "lastName": "Tang",
      "email": "jeffrey.tang80@gmail.com",
      "phone": null,
      "netID": null,
      "UIN": null,
      "major": null,
      "birthdate": null,
      "github": null,
      "snapchat": null,
      "instagram": null,
      "areDuesPaid": null,
      "gradYear": null,
      "gradSemester": null,
      "classStanding": null,
      "generationYear": null,
      "generationSemester": null,
      "location": null,
      "role": null,
      "level": "DIRECTOR",
      "status": null,
      "_id": "XXXXX"
    }
  ]
}
```

**Request**
GET `/api/members/:memberId` Returns member info for specific member. Like GET `/api/members` but returns an object instead of a list of objects

**Request**
GET `/api/members/current` Returns current logged-in member info. `result` is null if not logged-in. Includes `areDuesPaid` and `level` for current user.

**Request**
POST `/api/members` Create new member. In request body, specify any field in Member to populate. Returns the created member with all fields, including its `id`. Requires director level

**Request**
GET `/api/members/:memberId/permissions` Gets fields that current user is allowed to view or edit for this member.

**Response**
HTTP/1.1 200 OK
```
{
  success: true,
  result: {
      view: [
        "firstName",
        "lastName",
        "email",
        "phone",
        "netID",
        "UIN",
        "major",
        "birthdate",
        "github",
        "snapchat",
        "instagram",
        "areDuesPaid",
        "gradYear",
        "gradSemester",
        "classStanding",
        "generationYear",
        "generationSemester",
        "location",
        "role",
        "level",
        "status",
        "_id"
      ],
      edit: [
        "firstName",
        "lastName",
        "email",
        "phone",
        "netID",
        "UIN",
        "major",
        "birthdate",
        "github",
        "snapchat",
        "instagram",
        "areDuesPaid",
        "gradYear",
        "gradSemester",
        "classStanding",
        "generationYear",
        "generationSemester",
        "location",
        "role",
        "level",
        "status"
      ]
    }
}
```

**Request**
PUT `/api/members/:memberId` Update fields for member. Specify fields in body in the same format as `result` field of GET `/api/members/:memberId`. Requires director level

Closes #2, closes #4, closes #5

### CODE REVIEW

File: api/src/utils/user-utils.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  fields.forEach((field) => {
+    delete objCopy[field];
+  });
+
+  return objCopy;
+};
+
+const filterSensitiveInfo = (user, additionalFields = []) => {
+  const omitFields = ['_id', 'oauthID', '__v', ...additionalFields];
```
Comment (mattwalo32) [inline api/src/utils/user-utils.js:12]: Could you put a TODO here to add 'notes' to the omit fields once we add them to the DB. If different people are working on this later, I could see that being overlooked

File: api/src/api/members.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  next();
+};
+
+router.get(
+  '/current',
+  errorWrap(async (req, res) => {
+    res.json({
+      success: true,
+      result: req.user ? filterSensitiveInfo(req.user) : null,
```
Comment (mattwalo32) [inline api/src/api/members.js:28]: Is this for getting the information of the current user? If so, don't we need to fetch the member data from the DB by oauthID?

File: api/src/api/members.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    if (!member) {
+      return res.status(404).json({
+        success: false,
+        message: 'member not found with id',
+      });
+    }
+
+    res.json({
+      success: true,
+      result: filterSensitiveInfo(member),
```
Comment (mattwalo32) [inline api/src/api/members.js:84]: Omit 'levels' and 'areDuesPaid' if not director

**Review by mattwalo32 — CHANGES_REQUESTED** [review]

File: api/src/api/members.js (hunk 2 above)
Comment (mattwalo32) [reply api/src/api/members.js:28]: Nevermind, I figured it out

File: api/src/utils/user-utils.js (hunk 1 above)
Comment (Qrtn) [reply api/src/utils/user-utils.js:12]: Yes, good point!

**Comment by Qrtn** [comment]
Thanks @mattwalo32 !!

File: api/src/api/members.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    res.json({
+      success: true,
+      result: filterViewableFields(req.user, member),
+    });
+  }),
+);
+
+router.get(
+  '/:memberId/permissions',
```
Comment (mattwalo32) [inline api/src/api/members.js:95]: Love this

File: api/src/api/members.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      },
+    });
+  }),
+);
+
+router.put(
+  '/:memberId',
+  requireRegistered,
+  validateMemberQuery,
+  errorWrap(async (req, res) => {
```
Comment (mattwalo32) [inline api/src/api/members.js:112]: We should do a check here to see if either the member is editing their own information or the member is a director.

**Review by mattwalo32 — CHANGES_REQUESTED** [review]
Looks great

File: api/src/api/members.js (hunk 5 above)
Comment (Qrtn) [reply api/src/api/members.js:112]: `validateMemberQuery` should take care of this by ensuring that each field is an editable field, as determined by `getEditableFields`, which takes into account the current user and the user being edited. So if current user is a member and trying to edit another user, they would have no editable fields and get a 401.

**Review by mattwalo32 — APPROVED** [review]

**Review by alicesf2 — APPROVED** [review]
Awesome work!!

**Pull request merged**